### DISCUSSION TITLE
[Feature] Support podman on FreeBSD

### DISCUSSION THREAD

**Discussion by Subito** [immich-app/immich#13670] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

immich works fine on podman running on FreeBSD, however I need to modify `start.sh` to explicitly ` export LD_LIBRARY_PATH` because somehow this env-var gets stripped from the container upon startup - even though it shows up in `podman inspect` correctly.

I suspect this has something to do with the way podman works on FreeBSD, so I would be very grateful if this workaround would be present upstream. Right now I build my own container with a modified `start.sh`.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by Math-Aumont** [comment]
Would it be possible to share how you configured podman to launch immich on FreeBSD ? 

I must be missing a step because podman always tries to retrieve an image created for FreeBSD and encounters an error. And I have activated the Linux services.

```
Trying to pull ghcr.io/immich-app/immich-machine-learning:release...
Error: internal error: unable to copy from source docker://ghcr.io/immich-app/immich-machine-learning:release: choosing an image from manifest list docker://ghcr.io/immich-app/immich-machine-learning:release: no i
mage found in image index for architecture "amd64", variant "", OS "freebsd"
Error: no container with name or ID "immich_machine_learning" found: no such container
Trying to pull docker.io/valkey/valkey@sha256:facc1d2c3462975c34e10fccb167bfa92b0e0dbd992fc282c29a61c3243afb11...
Error: internal error: unable to copy from source docker://valkey/valkey@sha256:facc1d2c3462975c34e10fccb167bfa92b0e0dbd992fc282c29a61c3243afb11: choosing an image from manifest list docker://valkey/valkey@sha256:
facc1d2c3462975c34e10fccb167bfa92b0e0dbd992fc282c29a61c3243afb11: no image found in manifest list for architecture "amd64", variant "", OS "freebsd"                                  
```

**Comment by torndar** [comment]
I'm also curious how @Subito got it working. I used the command `sudo podman-compose --verbose --podman-pull-args="--os=linux" --podman-build-args="--os=linux" --podman-run-args="--os=linux" up
` to force podman to pull the linux images but I get startup errors on the containers e.g. for `ghcr.io/immich-app/postgres` I get 
```
FATAL:  XX000: signalfd() failed
66d1f1f21d10 LOCATION:  InitializeLatchSupport, latch.c:294
66d1f1f21d10 child process exited with exit code 1
```

**Comment by Subito** [comment]
Honestly, I gave up. I had it running with a custom-build postgres jail which had the neccessary extensions in it to circumvent the issue with the postgres-container. Other than that, I ran it exactly as you described, @torndar. 

But for some strange reason, immich would freeze from time to time and require a restart running with podman. I tried a few times to debug and fix the issue, but unfortunately nobody was willing to invest time and energy on this quest and I couldn't figure it out by myself.

My solution in the end: A Linux-VM with docker.

If one of you figures this out and gets it running reliably, I would be thrilled.

**Comment by Subito** [comment]
Sorry, I just noticed (its been a while) - I had to fix `start.sh` to get it running. Here are my steps, in case someone wants to dig further:

`Containerfile`:
```
FROM ghcr.io/immich-app/immich-server:release

WORKDIR /usr/src/app
COPY start.sh .
```

`start.sh`:
```
#!/usr/bin/env bash

echo "Initializing Immich v1.119.1 with custum start.sh"

lib_path="/usr/lib/x86_64-linux-gnu/libmimalloc.so.2"
export LD_PRELOAD=""

read_file_and_export() {
if [ -n "" ]; then
content=""
export ""=""
unset ""
fi
}
read_file_and_export "DB_URL_FILE" "DB_URL"
read_file_and_export "DB_HOSTNAME_FILE" "DB_HOSTNAME"
read_file_and_export "DB_DATABASE_NAME_FILE" "DB_DATABASE_NAME"
read_file_and_export "DB_USERNAME_FILE" "DB_USERNAME"
read_file_and_export "DB_PASSWORD_FILE" "DB_PASSWORD"
read_file_and_export "REDIS_PASSWORD_FILE" "REDIS_PASSWORD"

export CPU_CORES="12"
echo "Detected CPU Cores: 12"
if [ "12" -gt 4 ]; then
export UV_THREADPOOL_SIZE=12
fi
export LD_LIBRARY_PATH=/usr/lib/jellyfin-ffmpeg/lib

exec node /usr/src/app/dist/main ""
```
(the important part is `export LD_LIBRARY_PATH`)

Than I build the custom container with: `sudo buildah build --no-cache --os=linux -t immich-server-custom:release .` and start immich:

```
sudo podman run --rm -d --os=linux \
	-v /data/containers/immich/library:/usr/src/app/upload \
	-p 3003:3003 \
	-e TZ="Europe/Berlin" \
	-e DB_HOSTNAME=postgres \
	-e DB_USERNAME=immich \
	-e DB_PASSWORD=password \
	-e DB_DATABASE_NAME=immich \
	-e DB_VECTOR_EXTENSION=pgvector \
	-e REDIS_HOSTNAME=redis \
	-e REDIS_DBINDEX=1 \
	-e REDIS_PASSWORD=password \
	-e LD_LIBRARY_PATH=/usr/lib/jellyfin-ffmpeg/lib \
	-v immich_model-cache:/cache \
	ghcr.io/immich-app/immich-machine-learning:release
sudo podman run --name=immich --restart=always -d --os=linux \
	-v /data/containers/immich/library:/usr/src/app/upload \
	-p 2283:2283 \
	-e TZ="Europe/Berlin" \
	-e DB_HOSTNAME=postgres \
	-e DB_USERNAME=immich \
	-e DB_PASSWORD=password \
	-e DB_DATABASE_NAME=immich \
	-e DB_VECTOR_EXTENSION=pgvector \
	-e REDIS_HOSTNAME=redis \
	-e REDIS_DBINDEX=1 \
	-e REDIS_PASSWORD=password \
	-e LD_LIBRARY_PATH=/usr/lib/jellyfin-ffmpeg/lib \
	--ulimit=nofile=-1 \
	localhost/immich-server-custom:release
```

I ran into networking trouble with podman, thats why I'm setting the `DB_HOSTNAME` and `REDIS_HOSTNAME` explicitly.

I have no idea if this approach still works. But at the time thats how I got it running.
